### PR TITLE
Bamboo e2e for job,job build,plan,plan build

### PR DESCRIPTION
**Pull request by mappjzc**
### Summary
Add Bamboo e2e test for job,job build,plan,plan build.
Add CollectDeploy
Add ExtractDeploy

### Does this close any open issues?
relate #3322 

### Screenshots
image
image
image
image

### CODE REVIEW

File: backend/plugins/bamboo/e2e/job_build_test.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+	// SELECT * FROM _raw_bamboo_api_job_build INTO OUTFILE "/tmp/_raw_bamboo_api_job_build.csv" FIELDS TERMINATED BY ',' OPTIONALLY ENCLOSED BY '"' LINES TERMINATED BY '\r\n';
+	dataflowTester.ImportCsvIntoRawTable("./raw_tables/_raw_bamboo_api_job_build.csv", "_raw_bamboo_api_job_build")
+
+	// verify extraction
+	dataflowTester.FlushTabler(&models.BambooJobBuild{})
+	dataflowTester.Subtask(tasks.ExtractJobBuildMeta, taskData)
+	dataflowTester.VerifyTable(
+		models.BambooJobBuild{},
+		"./snapshot_tables/_tool_bamboo_job_builds.csv",
+		e2ehelper.ColumnWithRawData(
```
Comment (warren830) [inline backend/plugins/bamboo/e2e/job_build_test.go:56]: Please use dataflowTester.VerifyTableWithOptions

File: backend/plugins/bamboo/e2e/job_test.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+	// SELECT * FROM _raw_bamboo_api_job INTO OUTFILE "/tmp/_raw_bamboo_api_job.csv" FIELDS TERMINATED BY ',' OPTIONALLY ENCLOSED BY '"' LINES TERMINATED BY '\r\n';
+	dataflowTester.ImportCsvIntoRawTable("./raw_tables/_raw_bamboo_api_job.csv", "_raw_bamboo_api_job")
+
+	// verify extraction
+	dataflowTester.FlushTabler(&models.BambooJob{})
+	dataflowTester.Subtask(tasks.ExtractJobMeta, taskData)
+	dataflowTester.VerifyTable(
+		models.BambooJob{},
+		"./snapshot_tables/_tool_bamboo_jobs.csv",
+		e2ehelper.ColumnWithRawData(
```
Comment (warren830) [inline backend/plugins/bamboo/e2e/job_test.go:54]: same as above

File: backend/plugins/bamboo/e2e/plan_build_test.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+	// SELECT * FROM _raw_bamboo_api_job_build INTO OUTFILE "/tmp/_raw_bamboo_api_job_build.csv" FIELDS TERMINATED BY ',' OPTIONALLY ENCLOSED BY '"' LINES TERMINATED BY '\r\n';
+	dataflowTester.ImportCsvIntoRawTable("./raw_tables/_raw_bamboo_api_plan_build.csv", "_raw_bamboo_api_plan_build")
+
+	// verify extraction
+	dataflowTester.FlushTabler(&models.BambooPlanBuild{})
+	dataflowTester.Subtask(tasks.ExtractPlanBuildMeta, taskData)
+	dataflowTester.VerifyTable(
+		models.BambooPlanBuild{},
+		"./snapshot_tables/_tool_bamboo_plan_builds.csv",
+		e2ehelper.ColumnWithRawData(
```
Comment (warren830) [inline backend/plugins/bamboo/e2e/plan_build_test.go:54]: same as above

File: backend/plugins/bamboo/e2e/plan_test.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+	// import raw data table
+	dataflowTester.ImportCsvIntoRawTable("./raw_tables/_raw_bamboo_api_plan.csv", "_raw_bamboo_api_plan")
+
+	// verify extraction
+	dataflowTester.FlushTabler(&models.BambooPlan{})
+	dataflowTester.Subtask(tasks.ExtractPlanMeta, taskData)
+	dataflowTester.VerifyTable(
+		models.BambooPlan{},
+		"./snapshot_tables/_tool_bamboo_plans.csv",
+		e2ehelper.ColumnWithRawData(
```
Comment (warren830) [inline backend/plugins/bamboo/e2e/plan_test.go:50]: same as above

File: backend/plugins/bamboo/e2e/raw_tables/_raw_bamboo_api_job.csv [hunk 5]
```diff
@@ -0,0 +1,5 @@
+"id","params","data","url","input","created_at"
```
Comment (warren830) [inline backend/plugins/bamboo/e2e/raw_tables/_raw_bamboo_api_job.csv:1]: 10 records are enough

File: backend/plugins/bamboo/e2e/raw_tables/_raw_bamboo_api_plan_build.csv [hunk 6]
```diff
@@ -0,0 +1,98 @@
+"id","params","data","url","input","created_at"
```
Comment (warren830) [inline backend/plugins/bamboo/e2e/raw_tables/_raw_bamboo_api_plan_build.csv:1]: 10 records are enough

File: backend/plugins/bamboo/e2e/snapshot_tables/_tool_bamboo_job_builds.csv [hunk 7]
```diff
@@ -0,0 +1,97 @@
+connection_id,job_build_key,job_key,plan_build_key,expand,number,build_number,job_name,plan_name,plan_key,project_name,project_key,build_result_key,life_cycle_state,build_started_time,pretty_build_started_time,build_completed_time,build_completed_date,pretty_build_completed_time,build_duration_in_seconds,build_duration,build_duration_description,build_relative_time,vcs_revision_key,build_test_summary,successful_test_count,failed_test_count,quarantined_test_count,skipped_test_count,continuable,once_off,restartable,not_run_yet,build_reason,reason_summary,state,build_state,job_result_key,_raw_data_params,_raw_data_table,_raw_data_id,_raw_data_remark
```
Comment (warren830) [inline backend/plugins/bamboo/e2e/snapshot_tables/_tool_bamboo_job_builds.csv:1]: 10 records are enough

File: backend/plugins/bamboo/models/deploy.go [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+*/
+
+package models
+
+import "github.com/apache/incubator-devlake/core/models/common"
+
+type BambooDeployEnvironment struct {
+	ConnectionId        uint64 `gorm:"primaryKey"`
+	ID                  uint64 `json:"id" gorm:"primaryKey"`
+	Key                 string `json:"key" gorm:"index;type:varchar(255)"`
```
Comment (warren830) [inline backend/plugins/bamboo/models/deploy.go:25]: please do not use `key`, use EnvKey to replace it

File: backend/plugins/bamboo/e2e/raw_tables/_raw_bamboo_api_job.csv (hunk 5 above)
Comment (mappjzc) [reply backend/plugins/bamboo/e2e/raw_tables/_raw_bamboo_api_job.csv:1]: This data is export from database, and has cropped already.

File: backend/plugins/bamboo/e2e/job_build_test.go (hunk 1 above)
Comment (mappjzc) [reply backend/plugins/bamboo/e2e/job_build_test.go:56]: There are 187 e2e test use VerifyTable but 59 e2e test use VerifyTableWithOptions.
we would better to fix all of it in the same pr.

**Review by likyh — APPROVED** [review]
I also think the raw data is too long but too simple. Such as [_tool_bamboo_job_builds.csv](https://github.com/apache/incubator-devlake/pull/4490/files#diff-ea103aa06daa136be288b7b0d9a64897d2ac0631929a585e98e7878ae224a342) have 96 rows but only 2 states (Successful/Failed).    
We can merge it and I think we should change them in next PR.

**Pull request merged**